card/mod/01_core/set/all/rules.rb

```diff
@@ -72,26 +130,155 @@ def toggle val
+  end
+  
   def rule_cache
     Card.cache.read('RULES') || begin        
-      hash = {}
-      ActiveRecord::Base.connection.select_all( Card::Set::All::Rules::RuleSQL ).each do |row|
-        setting_code = Card::Codename[ row['setting_id'].to_i ] or next
-        anchor_id = row['anchor_id']
-        set_class_id = anchor_id.nil? ? row['set_id'] : row['set_tag_id']
-    
-        set_class_code = Card::Codename[ set_class_id.to_i ] or next
-        hash_key = [ anchor_id, set_class_code, setting_code ].compact.map( &:to_s ) * '+'
-        hash[ hash_key ] = row['rule_id'].to_i
+      rule_hash = {}
+      all_rule_keys_with_id do |key,rule_id|
+        rule_hash[key] = rule_id
       end
-      Card.cache.write 'RULES', hash
+      
+      user_ids_hash = {}
+      rule_keys_hash = {}
+      all_user_rule_keys_with_id_and_user_id do |key, rule_id, user_id|
+        rule_hash[ user_rule_key(key,user_id) ] = rule_id
+        user_ids_hash[key] ||= []
+        user_ids_hash[key] << user_id
+        rule_keys_hash[user_id] ||= []
+        rule_keys_hash[user_id] << key
+      end
+      write_user_ids_cache user_ids_hash
+      write_rule_keys_cache rule_keys_hash
+      write_rule_cache rule_hash
+    end
+  end
+  
+  def user_rule_key key, user_id
+    "#{key}+#{user_id}"
+  end
+  
+  # all users that have a user-specific rule for a given rule key
+  def user_ids_cache
+    Card.cache.read('USER_IDS') || begin
+      rule_cache
+      Card.cache.read('USER_IDS')
+    end
+  end
+  
+  # all keys of user-specific rules for a given user
+  def rule_keys_cache
+    Card.cache.read('RULE_KEYS') || begin
+      rule_cache
+      Card.cache.read('RULE_KEYS')
     end
   end
   
   def clear_rule_cache
-    Card.cache.write 'RULES', nil
+    write_rule_cache nil
+    write_user_ids_cache nil
+    write_rule_keys_cache nil
+  end
+  
+  def clear_user_rule_cache
+    clear_rule_cache
+  end
+  
+  def refresh_rule_cache_for_user user_id
+    rule_hash = rule_cache
+    user_ids_hash = user_ids_cache
+    rule_keys_hash = rule_keys_cache
+    
+    cached_rule_keys_for(user_id).each do |key|
+      rule_hash[ user_rule_key(key, user_id) ] = nil
+      user_ids_hash[ key ].delete(user_id)
+    end
+    rule_keys_hash[ user_id ] = nil
+    
+    all_rule_keys_with_id_for(user_id) do |key, rule_id|
+      rule_hash[ user_rule_key(key,user_id) ] = rule_id
+      
+      user_ids_hash[ key ]      ||= []
+      user_ids_hash[ key ]      << user_id
+      rule_keys_hash[ user_id ] ||= []
+      rule_keys_hash[ user_id ] << key
+    end
+    write_user_ids_cache user_ids_hash
+    write_rule_keys_cache rule_keys_hash
+    write_rule_cache rule_hash
   end
   
+  def write_rule_cache hash
+    Card.cache.write 'RULES', hash
+  end
+  
+  def write_user_ids_cache hash
+    Card.cache.write 'USER_IDS', hash
+  end
+  
+  def write_rule_keys_cache hash
+    Card.cache.write 'RULE_KEYS', hash
+  end
+  
+  
+  
```

Comment:
man, this is a chunk of work!

I feel like I should write up comments for it to be sure that I understand it.  But I can't spot any problems with a quick look, and despite all the new functionality, if anything the code is more handsome than what it replaces.